### DISCUSSION TITLE
Documenting parameters in taxonomy

### DISCUSSION THREAD

**Discussion by ryanmwhitephd** [NCSLI-MII/measurand-taxonomy#52] [Ideas]
Parameters can represent various types of information and how the taxonomy documents that information is not clear at this time. For example, Parameters may refer to Unit under test (UUT) inputs, UUT input quantities, influence quantities that need to be a measured independent of the UUT (environmental conditions), parameter values taken as constants (assumed constant values rather than measured parameters). 

- Influence parameter can be described by its aspect or kind of quantity alone (e.g. Temperature)
- Influence parameter needs to be qualified (e.g. Temperature.Ambient)
- Influence parameter is described by another measurand (e.g. Measure.Temperature.Ambient)
- Parameter is a quantity or nominal property
- ...

This is kind of a blocker, see issue #44.
Let's brainstorm possible types of parameters and decide how to document and what to reference.

**Comment by MarkKuster** [comment]
I think we should control this in the schema or UIs to the extent effective. In general, parameters should refer to another taxon for the least ambiguity. We should discuss whether in those cases it should include the Measure-Source token or not. Another case: property values. Given that we want to limit the values to controlled lists (e.g., thermocouple types) we should control the parameter name also, e.g. ThermocoupleType. So maybe all parameter names conform either to another taxon or an entry from a separate controlled list. Perhaps we even change the schema to recognize two parameter types, e.g., quantities and nominal properties.

**Reply by SchwartzMichael** [reply]
The names will be controlled in the Taxon definition in the XML file. 

And we can expand the schema to include references/links to other related elements like MLayer, quantity kind, ISO<something>, SI units, etc. 

Including a cross-reference guide in the taxon definition will increase the user value of a taxon.

**Comment by ryanmwhitephd** [comment]
@colinindy, what about frequency? Can you describe this in relation to a measurand definition?

**Comment by SchwartzMichael** [comment]
I started using UUT.<Something> in Metrology.NET to it was a parameter that applied to the UUT only.
In Metrology.NET, the reference standards and the UUT come together, making the test ready for automation. 
An example would be the word "Range" a range parameter would be ambiguous (UUT range or Reference Standard Range). 
By adding UUT.Range separates it from the reference standard range, keeping range out of any parameter that is passed to the uncertainty calculators and/or SoA calculators. 

This was extended to UUT.Input to denote automatic connection messages for the Automation Software. 

Keep in mind I'm looking at this from the DCR point of view.  I want to digitally define the test requirements for a device I'm sending to a calibration lab. I want to define it in such detail that the calibration tech doesn't need a pre-written manual to perform the calibration.  So my use case is a little different. 

Note:  None of the Metrology.NET optional parameter extensions are needed in the standard configuration.  
This is the power of this format; any user of the taxonomy can add an optional parameter to further define the data.

**Comment by SchwartzMichael** [comment]
I think we should keep the parameter names short and focused around the Taxon's needs. 

We can also include data about the parameter in the taxon definition like the ID of the M-Layer definition and/or information about another taxon.  

This additional data should be in the Taxon Definition, and NOT part of the package data moved from system to system. 

We need to update the Taxon Schema to support the M-Layer, Quantity Kind, and Referenced Taxons for the parameters.

**Comment by ryanmwhitephd** [comment]
@MarkKuster, @SchwartzMichael 
We had some discussion yesterday. I think the schema changes are good. 

The Result seems appropiate, it describes the "resulting aspect of measurand Y is X", where X references the mLayer entity by identifier and name and Y is the measurand taxon name.
In a lot of cases, the required parameters would also include the same aspect as the result, e.g. Measure.Voltage the Result is Voltage and the first (required) parameter is Voltage. However, for other quantities there may be differences. In particular, once the mlayer addresses ratio and coefficient quantities (several examples of photometric quantities) differences between the required parameters and the Result aspect will arise which help to disambiguate the measurands further.
One example is (pressure) sensitivity (see #31). The Result should have an aspect for Sensitivity, whereas, the parameters should have required parameters for Pressure and Voltage. 

Documenting some examples would be useful. Otherwise, if we all agree, I can probably produce an updated version of the taxonomy conforming to the new schema using [Measurand Flask App](https://github.com/NCSLI-MII/measurand-flask-app) with mlayer references for the Result and Parameters.

**Comment by colinindy** [comment]
what if we simplify this by replacing the Result tag with a set of Parameter _Types_? 

**Property**: An enum factor that affects the measurement indirectly: e.g. Fluid {Oil, Water}, Reference material {Water, Air}, etc
**Influence**: An indirect _quantity_ that impacts the measurement: e.g. Frequency, Temperature
**Input**: A quantity that is part of the measurement function itself. e.g. For Measure.Ratio.Humidity.Relative, the measurand may have inputs of Temperature and Temperature.Dewpoint. For Measure.Temperature.Simulated.Thermocouple, Voltage is an input.
**Result**: The result of the measurement, the quantity of the measurand itself.

**Comment by colinindy** [comment]
@SchwartzMichael copying from #69 

Let me copy my proposed definitions here:

**Property**: An enum factor that affects the measurement indirectly: e.g. Fluid {Oil, Water}, Reference material {Water, Air}, etc
**Influence**: An indirect _quantity_ that impacts the measurement: e.g. Frequency, Temperature, Impedance
**Input**: A quantity that is part of the measurement function itself. e.g. For Measure.Ratio.Humidity.Relative, the measurand may have inputs of Temperature and Temperature.Dewpoint. For Measure.Temperature.Simulated.Thermocouple, Voltage is an input.
**Result**: The result of the measurement, the quantity of the measurand itself.

In your Impedance example, it'd be fine to define that with an Optional Parameter with a Type of "Influence". Europe is welcome to use it. It's not an Input. It's not a Result, and it's not a Property.

In my implementation, I need Influence, Input, and Output quantities defined. I also need any properties defined as such. This data is used in our system when configuring Measurands for different measurement groups within a procedure and specification. I think we have enough information to differentiate these up front. Some hands on work on our end with these show that a loose list of parameters doesn't really provide value.

**Comment by ryanmwhitephd** [comment]
Keep in mind, the taxon is generic and seperate from the instrument, procedure or capability. The indications from a particular instrument may have known influences that are different from another instrument, but they are both used for the same measurand. Ideally, the taxon definition separates theses details, leaving parameters that are common to the measurand but not to the instruments, and leaves the specifics to the user. 

The schema should support controlled lists, but the ones defined in the schema should be relatively few as well as nominal properties of taxons in the catalog. Again, leaving most of the details to the user but still allowing for valid taxonomy definitions.

The parameter is either a quantity (either an mlayer aspect or a taxon) or a nominal property (indicated as list). I don't think this requires any further typing, the schema just needs to enforce this somehow. @SchwartzMichael is this already part of the schema?

The Result is an aspect.

**Reply by colinindy** [reply]
I'm aligned with the general nature of how we define measurands. We are treating measurands as instrument and procedure agnostic.

We have Result already broken out, and we can interrogate the definition to parse out the property parameters. The only missing piece is a differentiation between influence quantities and input quantities. If we're not differentiating these, why include parameters at all?

**Comment by ryanmwhitephd** [comment]
@colinindy more examples may help. Use the existing taxons posted in the issues, maybe we can sort this out with specific example.
The intent is to make the measurands exhaustive in terms of parameters, so that the optional parameters can be used at any metrology level. By referencing another taxon, this is likely the extent of the specificity. 
@MarkKuster may have some examples where an input in one case may be an influence in another. 

Consider gauge block length. This requires a known vacuum wavelength (we count the number of fringes in an interference pattern), the vacuum wavelength is an input (source of length). However, the drift of the calibrated vaccum wavelength is an uncertainty component. We measure the gauge block's temperature directly to correct for thermal expansion. Are these inputs, influences, or both?

**Comment by ryanmwhitephd** [comment]
The following guidance, posted by @MarkKuster in #88 should be incorporated into the documentation, along with more examples. However, some further technical issues to deal (see 2. below). Significant progress made in this discussion!

More thinking to do still but @MarkKuster suggestions at this point:

1. The measurement function calculates the measurement result from the input quantities, serves as the base for uncertainty calculations and traceability. The uncertainty-traceability model, which incorporates one or more measurands, will identify any input quantities. A given input quantity may come from the measurand in which it serves as an influence quantity (such as ambient temperature used as a correction) or it may get added strictly from the measurement process. The measurand stands independent of the measurement process. Therefore, the measurand should not designate any parameters as input quantities.
2. My previous inclination to identify parameters as other taxa probably doesn't work in general. I agree that the taxon-measurand domain may govern its parameter IDs. If we simply identify them by unique name within a given measurand, then we may match parameters in different instantiations of the same measurand-taxon. Two parameters of the same aspect or property in the same measurand should have different names. We shouldn't rely on the (human-readable) parameter definition for this.
3. The dew-point temperature exemplifies an input quantity that arises with a specific measurement process and therefore doesn't belong in the measurand definition.

**Comment by SchwartzMichael** [comment]
Having an external dataset with common names and lists would help maintain a standardized naming convention. 
However, I don't think we need an ID in the parameter elements.   

This would be more of a tool for when New Taxons are being created.  Once a Taxon is created, approved, and deployed.  The name and optional list will become part of that taxon definition until it is updated (approved and deployed).

**Comment by ryanmwhitephd** [comment]
The parameters of a taxon do not require an id, however, parameter data elements may require a reference id to validate that the external reference (controlled list, quantity reference, mlayer id) exists.

**Reply by ryanmwhitephd** [reply]
XML spec does not permit  validating an XML instance containing IDREF attribute values that refer to ID attribute values in a separate XML instance (somewhat obvious). Similar problem with RDF. Stick with xs:Name types for id elements.

**Comment by ryanmwhitephd** [comment]
W3C reference for modeling lists: https://www.w3.org/2002/ws/databinding/examples/6/09/List/
